pkg/controllers/metrics/node/counter.go

```diff
@@ -0,0 +1,241 @@
+	archValues := knownValuesForNodeLabels[nodeLabelArch]
+	instanceTypeValues := knownValuesForNodeLabels[nodeLabelInstanceType]
+	osValues := knownValuesForNodeLabels[nodeLabelOS]
+	zoneValues := knownValuesForNodeLabels[nodeLabelZone]
+
+	errors := make([]error, 0, len(archValues)*len(instanceTypeValues)*len(osValues)*len(zoneValues))
+
+	// 1. Publish the count of all nodes associated with `provisioner`.
+	nodeLabels := client.MatchingLabels{nodeLabelProvisioner: provisioner}
+	errors = append(errors, consumeNodesWith(nodeLabels, func(nodes []v1.Node) error {
+		return publishCount(nodeCountByProvisioner, metricLabelsFrom(nodeLabels), len(nodes))
+	}))
+
+	// 2. Publish the count of all nodes associated with `provisioner`, in `zone`, and reported as "ready".
+	// 3. Publish the count of all nodes with `arch`, associated with `provisioner`, in `zone`, and reported as "ready".
+	// 4. Publish the count of all nodes with `instanceType`, associated with `provisioner`, in `zone`, and reported as "ready"
+	// 5. Publish the count of all nodes with `os`, associated with `provisioner`, in `zone`, and reported as "ready".
+	for _, zone := range zoneValues {
+		nodeLabels = client.MatchingLabels{
+			nodeLabelProvisioner: provisioner,
+			nodeLabelZone:        zone,
+		}
+		errors = append(errors, consumeNodesWith(nodeLabels, filterReadyNodes(func(readyNodes []v1.Node) error {
+			return publishCount(readyNodeCountByProvisionerZone, metricLabelsFrom(nodeLabels), len(readyNodes))
+		})))
+
+		for _, arch := range archValues {
+			nodeLabels := client.MatchingLabels{
+				nodeLabelArch:        arch,
+				nodeLabelProvisioner: provisioner,
+				nodeLabelZone:        zone,
+			}
+			errors = append(errors, consumeNodesWith(nodeLabels, filterReadyNodes(func(readyNodes []v1.Node) error {
+				return publishCount(readyNodeCountByArchProvisionerZone, metricLabelsFrom(nodeLabels), len(readyNodes))
+			})))
+		}
+
+		for _, instanceType := range instanceTypeValues {
+			nodeLabels := client.MatchingLabels{
+				nodeLabelInstanceType: instanceType,
+				nodeLabelProvisioner:  provisioner,
+				nodeLabelZone:         zone,
+			}
+			errors = append(errors, consumeNodesWith(nodeLabels, filterReadyNodes(func(readyNodes []v1.Node) error {
+				return publishCount(readyNodeCountByInstancetypeProvisionerZone, metricLabelsFrom(nodeLabels), len(readyNodes))
+			})))
+		}
+
+		for _, os := range osValues {
+			nodeLabels := client.MatchingLabels{
+				nodeLabelOS:          os,
+				nodeLabelProvisioner: provisioner,
+				nodeLabelZone:        zone,
+			}
+			errors = append(errors, consumeNodesWith(nodeLabels, filterReadyNodes(func(readyNodes []v1.Node) error {
+				return publishCount(readyNodeCountByOsProvisionerZone, metricLabelsFrom(nodeLabels), len(readyNodes))
+			})))
+		}
+	}
+
+	// Combine will filter out `nil` values; if no errors remain then it will return `nil`.
```

Comment:
nit: I'd just remove this comment (if people are confused they can look up what `multierr.Combine` does, but it's a bit overkill to document it here.)